### DISCUSSION TITLE
Setter is not setting

### DISCUSSION THREAD

**Discussion by MPiland** [Automattic/mongoose#13982] [Q&A] [answered]
I cannot for the life of me understand what I'm doing wrong. I have a model that looks essentially like this

```
export const itemSchema = new Schema(
  {
...
damage: {
      diceCount: Number,
      diceValue: Number,
      diceShort: {
        type: String,
        set: () => {
          let diceShortValue = ''
          console.log('updated this=', this._update)
          if (this._update.damage.diceCount && this._update.damage.diceValue) {
            diceShortValue = `${this._update.damage.diceCount}d${this._update.damage.diceValue}`
          }
          return diceShortValue
        },
      }
...
},
  {
    toJSON: { setters: true, minimize: true, virtuals: true },
    toObject: { setters: true, minimize: true, virtuals: true },
    runSettersOnQuery: true,
  }
)

export default mongoose.model('Item', itemSchema)
```

My API for updating a document looks like

```
export const updateItem = async (req, res) => {
  const { _id } = req.body

  if (!_id) {
    return res.status(400).json({ message: 'Item ID is required' })
  }

  const updatedItem = req.body

  const item = await Item.findOneAndUpdate({ _id: _id }, updatedItem, {
    upsert: true,
    new: true,
  }).exec()

  if (!item) {
    return res.status(400).json({ message: 'Item not found' })
  }

  res.json(`'${item.value}' updated`)
}
```

I have confirmed that the setter is not firing no matter what I change on the document. I can't understand what I'm doing wrong here. Any help would be appreciated. This seems to be a difficult concept for me to grasp. I basically am trying to create the value for the key whenever the user updates/creates the document. If I'm going about this the wrong way, please tell me how I should be approaching this.

I'm using this in a MERN application with RTK Query if that helps.

**Comment by vkarpov15** [comment]
Can you print out what is the value of `updatedItem`?

**Reply by MPiland** [reply, marked as answer]
I ended up just reworking it to use virtuals instead. I guess setters can't access values outside of their own. The virtuals are fine, though, as it's not actually needed to be stored and just displayed.
